sdks/python/apache_beam/runners/dataflow/test_dataflow_runner.py

```diff
@@ -44,34 +44,41 @@ def run_pipeline(self, pipeline):
 
     self.result = super(TestDataflowRunner, self).run_pipeline(pipeline)
     if self.result.has_job:
-      project = pipeline._options.view_as(GoogleCloudOptions).project
-      region_id = pipeline._options.view_as(GoogleCloudOptions).region
-      job_id = self.result.job_id()
       # TODO(markflyhigh)(BEAM-1890): Use print since Nose dosen't show logs
       # in some cases.
-      print (
-          'Found: https://console.cloud.google.com/dataflow/jobsDetail'
-          '/locations/%s/jobs/%s?project=%s' % (region_id, job_id, project))
+      print('Found: %s.' % self.build_console_url(pipeline.options))
 
     if not options.view_as(StandardOptions).streaming:
       self.result.wait_until_finish()
     else:
-      # TODO: Ideally, we want to wait until workers start successfully.
-      self.wait_until_running()
+      self.wait_until_in_state(PipelineState.RUNNING)
 
     if on_success_matcher:
       from hamcrest import assert_that as hc_assert_that
       hc_assert_that(self.result, pickler.loads(on_success_matcher))
 
+    if options.view_as(StandardOptions).streaming:
+      self.result.cancel()
```

Comment:
Do you want to cancel the job immediately after it is reaching to running state? Should we wait a bit first?

Reply:
Because we want pipeline to work for a bit after the setup is completed. setup -> cancel does not really allow anytime for pipeline to process data.

Reply:
I see. A customized matcher is probably one approach for this. Like `PubSubMessageMatcher`, the verification is blocked until messaged pulled (or timeout). So wait logic inside `on_success_matcher` can give pipeline time to process the data. However, cancel will be called immediately after setup only if no `on_success_matcher` is provided.

Do we want a waiting step here as a general step of the test?

Reply:
But if you can cancel, before you start the verification process, the pipeline may still not be able to process the data. 

I think we can simply sleep a bit before calling cancel. (Also note that pipeline running state is reached before workers actually doing meaning work. They still need to install SDK etc.)

Reply:
I missed the success matchers above. This is fine as it is.